update readme (#1)
avo-hq/prop_initializer: README.md

# PropInitializer Gem Documentation

## Overview

> The `PropInitializer` gem is a flexible tool for defining properties on Ruby classes. This gem is an adaptation of the [Literal Gem](https://github.com/joeldrapper/literal), with custom modifications tailored to the needs of [Avo](https://avohq.io). We sincerely thank [Joel Drapper](https://github.com/joeldrapper) for the inspiration and base code that made this possible.

Reviewer: ```suggestion
The `prop_initializer` gem is a flexible tool for defining properties on Ruby classes.

It's a fork of the [Literal Gem](https://github.com/joeldrapper/literal), with a few tweaks. We sincerely thank [Joel Drapper](https://github.com/joeldrapper) for the inspiration and base code that made this possible.
```

> With `PropInitializer`, you can easily declare properties for any class, giving you flexible options for default values and more. However, unlike [Literal](https://github.com/joeldrapper/literal), `PropInitializer` simplifies by removing strict typing requirements, providing a more lightweight and adaptable interface.

Reviewer: ```suggestion
With `prop_initializer`, you can easily declare properties for any class, giving you flexible options for default values and more. However, the scope is narrowed down by removing strict typing requirements, providing a more lightweight and adaptable interface.
```

---

## Installation

> To use `PropInitializer` in your Ruby project, add the following to your Gemfile:

Reviewer: ```suggestion
To use `prop_initializer ` in your Ruby project, add the following to your Gemfile:
```

```ruby
gem "prop_initializer"
```

Then, run the bundle command to install it:

```bash
bundle install
```

## Usage

> To start using `PropInitializer`, you need to include the module in the class where you want to define properties.

Reviewer: ```suggestion
To start using `prop_initializer `, you need to include the module in the class where you want to define properties.
```

end
```

### Step 3: Accessing Properties

By default, `PropInitializer` generates a writer for each declared property. The properties are stored as instance variables, which can be accessed within the class:

```ruby
@name    # Accesses the value of the 'name' property
@size    # Accesses the value of the 'size' property
```

#### Public Reader:

If you want to generate a public reader (getter) for a property, use the `reader: :public` option when defining the property:

```ruby
prop :title, reader: :public
```

This will automatically generate a public getter method, allowing you to retrieve the property like so:

```ruby
my_class_instance.title  # Public getter for the 'title' property
```

### Example:

```ruby
class MyComponent
  extend PropInitializer::Properties

  prop :name, reader: :public
  prop :size, default: :md, reader: :public
  prop :args, kind: :*
  prop :kwargs, kind: :**
  prop :icon do |value|
    value&.to_sym
  end
end

component = MyComponent.new(name: "Button", size: :lg)
component.name # => "Button"
component.size # => :lg
```

---

## Key Differences from [Literal](https://github.com/joeldrapper/literal)

> While `PropInitializer` is based on the [Literal Gem](https://github.com/joeldrapper/literal), there are some important differences:

Reviewer: ```suggestion
While `prop_initializer` is based on the [Literal Gem](https://github.com/joeldrapper/literal), there are some important differences:
```

> - **No Type Requirement:** [Literal](https://github.com/joeldrapper/literal)'s properties system enforces types, but `PropInitializer` omits this for flexibility. You can define properties without needing to specify a type.

Reviewer: ```suggestion
- **No Type Requirement:** [Literal](https://github.com/joeldrapper/literal)'s properties system enforces types, while `prop_initializer` omits them for flexibility. You can define properties without needing to specify a type.
```

  
- **Simplified Initializer:** The initialization process has been modified to avoid requiring types at the time of property definition.

---

## Acknowledgements

Special thanks to the team at [Literal](https://github.com/joeldrapper/literal) for their pioneering work on property-based object initialization. This gem builds upon the foundation laid by their work.

If you're looking for a more type-strict approach to property initialization, we encourage you to check out the original [Literal Gem](https://github.com/joeldrapper/literal).

--- 

> Thank you for using `PropInitializer`! We hope this tool helps make your property management more efficient and adaptable in your Ruby projects.

Reviewer: ```suggestion
Thank you for using `prop_initializer`! We hope this tool helps make your property management more efficient and adaptable in your Ruby projects.
```
